cpp/src/arrow/compute/kernels/vector_sort.cc

```diff
@@ -1778,6 +1799,736 @@ class SortIndicesMetaFunction : public MetaFunction {
+      return Status::OK();
+    }
+    if (options_.k > table_.num_rows()) {
+      options_.k = table_.num_rows();
+    }
+    std::function<bool(const uint64_t&, const uint64_t&)> cmp;
+    SelectKComparator<sort_order> select_k_comparator;
+    cmp = [&](const uint64_t& left, const uint64_t& right) -> bool {
+      auto chunk_left = first_sort_key.GetChunk<ArrayType>(left);
+      auto chunk_right = first_sort_key.GetChunk<ArrayType>(right);
+      auto value_left = chunk_left.Value();
+      auto value_right = chunk_right.Value();
+      if (value_left == value_right) {
+        return comparator.Compare(left, right, 1);
+      }
+      return select_k_comparator(value_left, value_right);
+    };
+    arrow::internal::Heap<uint64_t, decltype(cmp)> heap(cmp);
+
+    std::vector<uint64_t> indices(num_rows);
+    uint64_t* indices_begin = indices.data();
+    uint64_t* indices_end = indices_begin + indices.size();
+    std::iota(indices_begin, indices_end, 0);
+
+    auto end_iter =
+        this->PartitionNullsInternal<InType>(indices_begin, indices_end, first_sort_key);
+    auto kth_begin = indices_begin + options_.k;
+
+    if (kth_begin > end_iter) {
+      kth_begin = end_iter;
+    }
+    uint64_t* iter = indices_begin;
+    for (; iter != kth_begin; ++iter) {
+      heap.Push(*iter);
+    }
+    for (; iter != end_iter && !heap.empty(); ++iter) {
+      uint64_t x_index = *iter;
+      uint64_t top_item = heap.top();
+      if (cmp(x_index, top_item)) {
+        heap.ReplaceTop(x_index);
+      }
+    }
+    if (options_.keep_duplicates == true) {
+      iter = indices_begin;
+      for (; iter != end_iter; ++iter) {
+        uint64_t x_index = *iter;
+        auto top_item = heap.top();
+        if (x_index != top_item) {
+          auto chunk_left = first_sort_key.GetChunk<ArrayType>(x_index);
+          auto chunk_right = first_sort_key.GetChunk<ArrayType>(top_item);
+          auto xval = chunk_left.Value();
+          auto top_value = chunk_right.Value();
+          if (xval == top_value && comparator.Equals(x_index, top_item, 1)) {
+            heap.Push(x_index);
+          }
+        }
+      }
+    }
+    int64_t out_size = static_cast<int64_t>(heap.size());
+    ARROW_ASSIGN_OR_RAISE(
+        auto take_indices,
+        MakeMutableArrayForFixedSizedType(uint64(), out_size, ctx_->memory_pool()));
+    auto* out_cbegin = take_indices->GetMutableValues<uint64_t>(1) + out_size - 1;
+    while (heap.size() > 0) {
+      *out_cbegin = heap.top();
+      heap.Pop();
+      --out_cbegin;
+    }
+    ARROW_ASSIGN_OR_RAISE(*output_, Take(Datum(table_), Datum(std::move(take_indices)),
+                                         TakeOptions::NoBoundsCheck(), ctx_));
+    return Status::OK();
+  }
+
+  ExecContext* ctx_;
+  const Table& table_;
+  SelectKOptions options_;
+  Datum* output_;
+  std::vector<ResolvedSortKey> sort_keys_;
+  Comparator comparator_;
+  Status status_;
+};
+
+template <SortOrder sort_order>
```

Comment:
Another thought: instead of a single SortOrder, why not accept SortKeys like the sort kernels? Then you could emulate a query like `... ORDER BY foo ASC, bar DESC LIMIT 5` which this kernel can't currently do.

Reply:
Sure, I create a new kernel called `Result select_k(datum, options)` where `options.sort_keys` type is `std::vector<SortKey>`.